core/runtime/src/main/java/io/quarkus/runtime/StartupContext.java

```diff
@@ -76,14 +84,14 @@ public void close() {
         lastShutdownTasks.clear();
     }
 
-    private void runAllInReverseOrder(List<Runnable> tasks) {
-        List<Runnable> toClose = new ArrayList<>(tasks);
-        Collections.reverse(toClose);
-        for (Runnable r : toClose) {
+    private void runAllInReverseOrder(Deque<Runnable> tasks) {
+        Deque<Runnable> toClose = new ArrayDeque<>(tasks);
```

Comment:
Hm, the `ConcurrentLinkedDeque` iterator is weakly consistent so I think that we don't need to create a new deque here but instead just leverage the `Deque#descendingIterator()`? The collection is cleared in the `StartupContext#close()` method anyway.

Reply:
> To protect myself from adding elements to ConcurrentLinkedDeque while reading the collection

Hm, I don't understand...

> In a general sense you are right of course, the only thing is that here you can use a regular iterator, because I already add handlers using the addFirst method.

Ok but my point is that you don't need to create a new `ArrayDeque` at all, so something like this should work:

```java
for (Runnable task :  tasks) {
   try {
      task.run();
   } catch (Throwable ex) {
      LOG.error("Running a shutdown task failed", ex);
   }
}
```

Reply:
Since it is a deque, you could skip the iteration altogether and say `while (!deque.isEmpty()) { var item = deque.removeLast(); ... }`.